pkg/router/controller/route_secret_manager_test.go

```diff
@@ -0,0 +1,1139 @@
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/openshift/library-go/pkg/route/secretmanager/fake"
+
+	routev1 "github.com/openshift/api/route/v1"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/watch"
+	testclient "k8s.io/client-go/kubernetes/fake"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type testSARCreator struct {
+	allow bool
+	err   error
+	sar   *authorizationv1.SubjectAccessReview
+}
+
+func (t *testSARCreator) Create(_ context.Context, subjectAccessReview *authorizationv1.SubjectAccessReview, _ metav1.CreateOptions) (*authorizationv1.SubjectAccessReview, error) {
+	t.sar = subjectAccessReview
+	return &authorizationv1.SubjectAccessReview{
+		Status: authorizationv1.SubjectAccessReviewStatus{
+			Allowed: t.allow,
+		},
+	}, t.err
+}
+
+type testSecretGetter struct {
+	namespace string
+	secret    *corev1.Secret
+}
+
+func (t *testSecretGetter) Secrets(_ string) corev1client.SecretInterface {
+	return testclient.NewSimpleClientset(t.secret).CoreV1().Secrets(t.namespace)
+}
+
+// fakeSecretInformer will list/watch only one secret inside a namespace
+func fakeSecretInformer(ctx context.Context, fakeKubeClient *testclient.Clientset, namespace, name string) cache.SharedInformer {
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", name).String()
+	klog.Info(fieldSelector)
+	return cache.NewSharedInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = fieldSelector
+				return fakeKubeClient.CoreV1().Secrets(namespace).List(ctx, options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = fieldSelector
+				return fakeKubeClient.CoreV1().Secrets(namespace).Watch(ctx, options)
+			},
+		},
+		&corev1.Secret{},
+		0,
+	)
+}
+
+func fakeSecret(namespace, name string, secretType corev1.SecretType, data map[string][]byte) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Data: data,
+		Type: secretType,
+	}
+}
+
+type fakePluginNew struct {
+	t      watch.EventType
+	route  *routev1.Route
+	err    error
+	stopCh chan struct{}
+}
+
+func (p *fakePluginNew) HandleRoute(t watch.EventType, route *routev1.Route) error {
+	defer close(p.stopCh)
+	p.t, p.route = t, route
+	return p.err
+}
+
+func (p *fakePluginNew) HandleNode(t watch.EventType, node *corev1.Node) error {
+	return fmt.Errorf("not expected")
+}
+
+func (p *fakePluginNew) HandleEndpoints(watch.EventType, *corev1.Endpoints) error {
+	return fmt.Errorf("not expected")
+}
+func (p *fakePluginNew) HandleNamespaces(namespaces sets.String) error {
+	return fmt.Errorf("not expected")
+}
```

Comment:
Does `fakePluginNew` implements some interface? Why do you need all these methods defined?

Reply:
Yes, it implements the Plugin interface.

https://github.com/openshift/router/blob/56ab14f257e123317c00ce7ba5b4ecae7b2e0102/pkg/router/interfaces.go#L13-L20

Reply:
Maybe a simple type check would be of help here:
```go
var _ Plugin = &fakePluginDone{}
```